File: src/field-watershed.ts
/**
 * Segmentation de limites de parcelles par watershed marqué, sur une image de
 * "force de frontière" combinant :
 *   - le gradient spatial du NDVI médian de la saison (les vraies limites de
 *     champs correspondent à des sauts brusques de NDVI)
 *   - l'écart-type temporel du NDVI sur plusieurs sous-périodes de la saison
 *     (deux parcelles adjacentes ont souvent des calendriers de culture
 *     différents, donc une variabilité temporelle différente, même quand
 *     elles se ressemblent à un instant donné)
 *
 * Aucun modèle de machine learning : uniquement des opérations GEE classiques
 * (gradient, reduce, stdDev) + un algorithme de watershed "priority-flood"
 * implémenté ici en pur TypeScript, + un traçage de contours (Moore-neighbor
 * tracing) et une simplification Douglas-Peucker.
 */

// ── Web Mercator (mètres) : conversion lat/lng <-> mètres EPSG:3857 ──
// (utilisé comme grille pour image:computePixels, car les distances y sont en
// mètres, contrairement à EPSG:4326 en degrés)
const EARTH_RADIUS_M = 6_378_137;

export function lngLatToMercatorMeters(lng: number, lat: number): { x: number; y: number } {
  const x = (lng * Math.PI * EARTH_RADIUS_M) / 180;
  const clampedLat = Math.min(Math.max(lat, -85.05112878), 85.05112878);
  const y = EARTH_RADIUS_M * Math.log(Math.tan(Math.PI / 4 + (clampedLat * Math.PI) / 360));
  return { x, y };
}

export function mercatorMetersToLngLat(x: number, y: number): { lng: number; lat: number } {
  const lng = (x / (Math.PI * EARTH_RADIUS_M)) * 180;
  const lat = ((2 * Math.atan(Math.exp(y / EARTH_RADIUS_M)) - Math.PI / 2) * 180) / Math.PI;
  return { lng, lat };
}

// ── Parsing minimal du format .npy (tableau 2D float32, C-order) ──

function readNpyHeader(buffer: ArrayBuffer): { headerText: string; dataStart: number } {
  const bytes = new Uint8Array(buffer);
  const magic = String.fromCharCode(...bytes.slice(1, 6));
  if (magic !== "NUMPY") throw new Error("Format .npy invalide (magic manquant).");

  const majorVersion = bytes[6];
  const headerLenBytes = majorVersion >= 2 ? 4 : 2;
  const headerLenOffset = 8;
  const headerLen = majorVersion >= 2
    ? new DataView(buffer, headerLenOffset, 4).getUint32(0, true)
    : new DataView(buffer, headerLenOffset, 2).getUint16(0, true);
  const headerStart = headerLenOffset + headerLenBytes;
  const headerText = String.fromCharCode(...bytes.slice(headerStart, headerStart + headerLen));
  return { headerText, dataStart: headerStart + headerLen };
}

export function parseNpyFloat32(buffer: ArrayBuffer): { data: Float32Array; shape: number[] } {
  const { headerText, dataStart } = readNpyHeader(buffer);

  const shapeMatch = headerText.match(/'shape':\s*\(([^)]*)\)/);
  const descrMatch = headerText.match(/'descr':\s*'([^']+)'/);
  if (!shapeMatch || !descrMatch) throw new Error("En-tête .npy illisible.");
  const shape = shapeMatch[1].split(",").map((s) => s.trim()).filter(Boolean).map(Number);
  const descr = descrMatch[1];
  if (!/f4$/.test(descr)) throw new Error(`Type .npy non supporté : ${descr} (float32 attendu).`);
  const littleEndian = !descr.startsWith(">");

  const count = shape.reduce((a, b) => a * b, 1);
  const view = new DataView(buffer, dataStart, count * 4);
  const data = new Float32Array(count);
  for (let i = 0; i < count; i++) data[i] = view.getFloat32(i * 4, littleEndian);
  return { data, shape };
}

/**
 * Parse un .npy "structuré" (dtype = liste de champs), tel que renvoyé par
 * image:computePixels quand plusieurs bandIds sont demandés en un seul appel :
 * chaque pixel est un enregistrement contenant une valeur par bande, dans l'ordre
 * du dtype — pas un tableau plan par bande (BSQ), mais entrelacé par pixel (BIP).
 * Accepte f4 (float32) et f8 (float64) par champ — GEE renvoie f8 pour un export
 * multi-bandes même quand les bandes sources sont castées en float32 côté serveur.
 * Retourne un Float32Array par bande, déjà dé-entrelacé (une bande = un tableau plat).
 */
export function parseNpyStructuredFloat32(buffer: ArrayBuffer): { bands: Record<string, Float32Array>; shape: number[] } {
  const { headerText, dataStart } = readNpyHeader(buffer);

  const shapeMatch = headerText.match(/'shape':\s*\(([^)]*)\)/);
  if (!shapeMatch) throw new Error("En-tête .npy illisible (shape manquant).");
  const shape = shapeMatch[1].split(",").map((s) => s.trim()).filter(Boolean).map(Number);

  const fieldMatches = [...headerText.matchAll(/\('([^']+)',\s*'([^']+)'\)/g)];
  if (fieldMatches.length === 0) throw new Error("En-tête .npy illisible (dtype structuré manquant).");

  let offset = 0;
  const fields = fieldMatches.map(([, name, descr]) => {
    const byteSize = descr.endsWith("f4") ? 4 : descr.endsWith("f8") ? 8 : null;
    if (byteSize === null) throw new Error(`Type .npy non supporté pour '${name}' : ${descr} (float32/float64 attendu).`);
    const field = { name, byteSize, littleEndian: !descr.startsWith(">"), offset };
    offset += byteSize;
    return field;
  });
  const recordSize = offset;

  const count = shape.reduce((a, b) => a * b, 1);
  const view = new DataView(buffer, dataStart, count * recordSize);

  const bands: Record<string, Float32Array> = {};
  for (const field of fields) bands[field.name] = new Float32Array(count);

  for (let i = 0; i < count; i++) {
    const recordOffset = i * recordSize;
    for (const field of fields) {
      const fieldOffset = recordOffset + field.offset;
      bands[field.name][i] = field.byteSize === 4
        ? view.getFloat32(fieldOffset, field.littleEndian)
        : view.getFloat64(fieldOffset, field.littleEndian);
    }
  }

  return { bands, shape };
}

/** Encode un tableau float32 (C-order) en bytes .npy — symétrique de parseNpyFloat32. */
export function encodeNpyFloat32(data: Float32Array, shape: number[]): Uint8Array {
  const magic = "\x93NUMPY";
  const version = new Uint8Array([1, 0]); // v1.0
  const dict = `{'descr': '<f4', 'fortran_order': False, 'shape': (${shape.join(", ")}${shape.length === 1 ? "," : ""}), }`;

  // L'en-tête total (magic + version + longueur + dict) doit être un multiple de 64 octets,
  // complété par des espaces puis un '\n' final — convention du format .npy.
  const unpaddedLen = magic.length + version.length + 2 + dict.length + 1;
  const padLen = (64 - (unpaddedLen % 64)) % 64;
  const header = dict + " ".repeat(padLen) + "\n";
  const headerLen = header.length;

  const buffer = new ArrayBuffer(magic.length + version.length + 2 + headerLen + data.length * 4);
  const bytes = new Uint8Array(buffer);
  let offset = 0;
  for (let i = 0; i < magic.length; i++) bytes[offset++] = magic.charCodeAt(i);
  bytes[offset++] = version[0];
  bytes[offset++] = version[1];
  new DataView(buffer, offset, 2).setUint16(0, headerLen, true);
  offset += 2;
  for (let i = 0; i < header.length; i++) bytes[offset++] = header.charCodeAt(i);

  const dataView = new DataView(buffer, offset, data.length * 4);
  for (let i = 0; i < data.length; i++) dataView.setFloat32(i * 4, data[i], true);

  return bytes;
}

// ── Watershed marqué (priority-flood / immersion à la Vincent-Soille) ──

export interface WatershedOptions {
  /** Percentile (0-1) sous lequel un pixel est candidat "germe" (fond de bassin). */
  seedPercentile?: number;
  /** Taille minimale (en pixels) d'une composante connexe pour être un germe valide. */
  minSeedPixels?: number;
}

class MinHeap<T> {
  private items: Array<{ priority: number; value: T }> = [];
  push(priority: number, value: T) {
    this.items.push({ priority, value });
    let i = this.items.length - 1;
    while (i > 0) {
      const parent = (i - 1) >> 1;
      if (this.items[parent].priority <= this.items[i].priority) break;
      [this.items[parent], this.items[i]] = [this.items[i], this.items[parent]];
      i = parent;
    }
  }
  pop(): T | undefined {
    if (this.items.length === 0) return undefined;
    const top = this.items[0];
    const last = this.items.pop()!;
    if (this.items.length > 0) {
      this.items[0] = last;
      let i = 0;
      for (;;) {
        const left = 2 * i + 1;
        const right = 2 * i + 2;
        let smallest = i;
        if (left < this.items.length && this.items[left].priority < this.items[smallest].priority) smallest = left;
        if (right < this.items.length && this.items[right].priority < this.items[smallest].priority) smallest = right;
        if (smallest === i) break;
        [this.items[smallest], this.items[i]] = [this.items[i], this.items[smallest]];
        i = smallest;
      }
    }
    return top.value;
  }
  get size() {
    return this.items.length;
  }
}

/**
 * @param strength Force de frontière par pixel (plus haut = plus proche d'une limite réelle).
 * @param barrier  true = pixel non cultivable (route, eau, bâti...) : jamais inondé, agit comme séparateur naturel.
 * @returns Un tableau de labels (0 = non affecté/barrière, >=1 = identifiant de parcelle).
 */
export function watershedSegment(
  strength: Float32Array,
  barrier: Uint8Array,
  width: number,
  height: number,
  options: WatershedOptions = {},
): Int32Array {
  const seedPercentile = options.seedPercentile ?? 0.2;
  const minSeedPixels = options.minSeedPixels ?? 6;

  const validValues: number[] = [];
  for (let i = 0; i < strength.length; i++) if (!barrier[i]) validValues.push(strength[i]);
  if (validValues.length === 0) return new Int32Array(strength.length);
  validValues.sort((a, b) => a - b);
  const threshold = validValues[Math.floor(validValues.length * seedPercentile)];

  const labels = new Int32Array(strength.length); // 0 = non labellisé
  const isSeedCandidate = (i: number) => !barrier[i] && strength[i] <= threshold;

  // Étiquetage des composantes connexes (4-connexité) parmi les candidats germes.
  let nextLabel = 0;
  const stack: number[] = [];
  for (let start = 0; start < strength.length; start++) {
    if (labels[start] !== 0 || !isSeedCandidate(start)) continue;
    nextLabel++;
    const componentPixels: number[] = [];
    stack.push(start);
    labels[start] = nextLabel;
    while (stack.length > 0) {
      const idx = stack.pop()!;
      componentPixels.push(idx);
      const x = idx % width;
      const y = (idx / width) | 0;
      const neighbors = [
        x > 0 ? idx - 1 : -1,
        x < width - 1 ? idx + 1 : -1,
        y > 0 ? idx - width : -1,
        y < height - 1 ? idx + width : -1,
      ];
      for (const n of neighbors) {
        if (n >= 0 && labels[n] === 0 && isSeedCandidate(n)) {
          labels[n] = nextLabel;
          stack.push(n);
        }
      }
    }
    if (componentPixels.length < minSeedPixels) {
      for (const idx of componentPixels) labels[idx] = 0;
      nextLabel--;
    }
  }

  // Inondation par priorité (priority-flood) depuis chaque germe.
  const heap = new MinHeap<number>();
  for (let idx = 0; idx < labels.length; idx++) {
    if (labels[idx] <= 0) continue;
    const x = idx % width;
    const y = (idx / width) | 0;
    const neighbors = [
      x > 0 ? idx - 1 : -1,
      x < width - 1 ? idx + 1 : -1,
      y > 0 ? idx - width : -1,
      y < height - 1 ? idx + width : -1,
    ];
    for (const n of neighbors) {
      if (n >= 0 && labels[n] === 0 && !barrier[n]) heap.push(strength[n], n);
    }
  }

  const visited = new Uint8Array(strength.length);
  while (heap.size > 0) {
    const idx = heap.pop()!;
    if (labels[idx] !== 0 || barrier[idx] || visited[idx]) continue;
    visited[idx] = 1;

    const x = idx % width;
    const y = (idx / width) | 0;
    const neighbors = [
      x > 0 ? idx - 1 : -1,
      x < width - 1 ? idx + 1 : -1,
      y > 0 ? idx - width : -1,
      y < height - 1 ? idx + width : -1,
    ];
    let assignedLabel = 0;
    for (const n of neighbors) {
      if (n >= 0 && labels[n] > 0) {
        assignedLabel = labels[n];
        break;
      }
    }
    if (assignedLabel === 0) continue;
    labels[idx] = assignedLabel;

    for (const n of neighbors) {
      if (n >= 0 && labels[n] === 0 && !barrier[n] && !visited[n]) heap.push(strength[n], n);
    }
  }

  return labels;
}

// ── Traçage de contours (Moore-neighbor tracing) + simplification Douglas-Peucker ──

/** Trace le contour extérieur de chaque label présent dans le tableau, en coordonnées pixels (coins de pixels). */
export function traceLabelContours(labels: Int32Array, width: number, height: number): Map<number, Array<{ x: number; y: number }>> {
  const contours = new Map<number, Array<{ x: number; y: number }>>();
  const visitedStart = new Set<string>();

  const at = (x: number, y: number) => (x < 0 || y < 0 || x >= width || y >= height ? 0 : labels[y * width + x]);

  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      const label = at(x, y);
      if (label <= 0) continue;
      // Pixel de bord gauche d'une région : premier pixel de la ligne appartenant à ce label,
      // ou pixel dont le voisin de gauche a un label différent.
      if (at(x - 1, y) === label) continue;
      const key = `${label}`;
      if (visitedStart.has(key)) continue; // un seul contour externe tracé par label (composante principale)
      visitedStart.add(key);

      const contour = mooreTrace(at, x, y, label, width, height);
      if (contour.length >= 3) contours.set(label, contour);
    }
  }
  return contours;
}

function mooreTrace(
  at: (x: number, y: number) => number,
  startX: number,
  startY: number,
  label: number,
  width: number,
  height: number,
): Array<{ x: number; y: number }> {
  // Directions en 8-connexité, sens horaire, en partant de l'Est (index 0).
  const directions = [
    [1, 0],   // 0 E
    [1, 1],   // 1 SE
    [0, 1],   // 2 S
    [-1, 1],  // 3 SW
    [-1, 0],  // 4 W
    [-1, -1], // 5 NW
    [0, -1],  // 6 N
    [1, -1],  // 7 NE
  ];
  const isForeground = (x: number, y: number) => at(x, y) === label;

  const boundary: Array<{ x: number; y: number }> = [{ x: startX, y: startY }];
  let cx = startX;
  let cy = startY;
  // On "arrive" virtuellement depuis l'Ouest (le pixel de départ est le plus à
  // gauche de sa ligne, donc son voisin Ouest est nécessairement fond).
  let backtrackDir = 4;
  const maxSteps = width * height * 8 + 8;
  let steps = 0;

  for (;;) {
    const searchStart = (backtrackDir + 1) % 8;
    let found = false;
    let nx = cx, ny = cy, newBacktrackDir = backtrackDir;
    for (let i = 0; i < 8; i++) {
      const d = (searchStart + i) % 8;
      const tx = cx + directions[d][0];
      const ty = cy + directions[d][1];
      if (isForeground(tx, ty)) {
        nx = tx;
        ny = ty;
        newBacktrackDir = (d + 4) % 8;
        found = true;
        break;
      }
    }
    if (!found) break; // pixel isolé (pas de voisin foreground)

    cx = nx;
    cy = ny;
    backtrackDir = newBacktrackDir;
    steps++;
    if (cx === startX && cy === startY) break; // retour au point de départ : contour bouclé
    boundary.push({ x: cx, y: cy });
    if (steps >= maxSteps) break;
  }

  return boundary;
}

export function simplifyPolygon(points: Array<{ x: number; y: number }>, epsilon: number): Array<{ x: number; y: number }> {
  if (points.length <= 2) return points;

  const perpendicularDistance = (p: { x: number; y: number }, a: { x: number; y: number }, b: { x: number; y: number }) => {
    const dx = b.x - a.x;
    const dy = b.y - a.y;
    const lengthSq = dx * dx + dy * dy;
    if (lengthSq === 0) return Math.hypot(p.x - a.x, p.y - a.y);
    const t = ((p.x - a.x) * dx + (p.y - a.y) * dy) / lengthSq;
    const projX = a.x + t * dx;
    const projY = a.y + t * dy;
    return Math.hypot(p.x - projX, p.y - projY);
  };

  const douglasPeucker = (pts: Array<{ x: number; y: number }>): Array<{ x: number; y: number }> => {
    if (pts.length <= 2) return pts;
    let maxDist = -1;
    let maxIndex = 0;
    for (let i = 1; i < pts.length - 1; i++) {
      const dist = perpendicularDistance(pts[i], pts[0], pts[pts.length - 1]);
      if (dist > maxDist) {
        maxDist = dist;
        maxIndex = i;
      }
    }
    if (maxDist > epsilon) {
      const left = douglasPeucker(pts.slice(0, maxIndex + 1));
      const right = douglasPeucker(pts.slice(maxIndex));
      return [...left.slice(0, -1), ...right];
    }
    return [pts[0], pts[pts.length - 1]];
  };

  return douglasPeucker(points);
}

File: app/services/field_watershed.py
"""Segmentation de limites de parcelles par watershed marqué, sur une image de
"force de frontière" combinant :
  - le gradient spatial du NDVI médian de la saison (les vraies limites de
    champs correspondent à des sauts brusques de NDVI)
  - l'écart-type temporel du NDVI sur plusieurs sous-périodes de la saison
    (deux parcelles adjacentes ont souvent des calendriers de culture
    différents, donc une variabilité temporelle différente, même quand
    elles se ressemblent à un instant donné)

Aucun modèle de machine learning : uniquement des opérations GEE classiques
(gradient, reduce, stdDev) + un algorithme de watershed "priority-flood"
implémenté ici en pur Python (port littéral du TypeScript original — la
priorité est la fidélité exacte de l'ordre d'inondation et du tracé de
contours sur les résultats affichés à l'utilisateur, PAS l'idiomaticité :
ne PAS remplacer par skimage.segmentation.watershed / cv2.findContours, dont
l'ordre de désambiguïsation des égalités diffère et produirait des formes de
parcelles légèrement différentes), + un traçage de contours (Moore-neighbor
tracing) et une simplification Douglas-Peucker.

Le parsing/encodage binaire .npy du fichier TS original (parseNpyFloat32,
parseNpyStructuredFloat32, encodeNpyFloat32) n'est PAS porté : numpy.load /
numpy.save gèrent nativement ce format (y compris les dtypes structurés
multi-champs), ce qui les rend inutiles côté Python.
"""

from __future__ import annotations

import io
import math

import numpy as np

EARTH_RADIUS_M = 6_378_137


def load_npy_band(content: bytes, band: str | None = None) -> tuple[np.ndarray, int, int]:
    """Charge un raster renvoyé par GEE image:computePixels au format NPY : un .npy "plat"
    (bandIds à un seul élément) ou "structuré" (bandIds à plusieurs éléments, un champ par
    bande, entrelacé par pixel) — numpy.load gère nativement les deux cas, contrairement au
    TS d'origine qui devait parser le format .npy à la main (parseNpyFloat32 /
    parseNpyStructuredFloat32, non portés : inutiles ici)."""
    arr = np.load(io.BytesIO(content), allow_pickle=False)
    if arr.dtype.names:
        field = band if band in arr.dtype.names else arr.dtype.names[0]
        values = arr[field].astype(np.float32)
    else:
        values = arr.astype(np.float32)
    height, width = values.shape[0], values.shape[1]
    return values.reshape(height, width), width, height


def load_npy_all_bands(content: bytes) -> tuple[dict[str, np.ndarray], int, int]:
    """Comme load_npy_band mais renvoie toutes les bandes d'un .npy structuré (une par
    champ nommé du dtype), chacune aplatie en 1D — équivalent de parseNpyStructuredFloat32."""
    arr = np.load(io.BytesIO(content), allow_pickle=False)
    if arr.dtype.names:
        height, width = arr.shape[0], arr.shape[1]
        bands = {name: arr[name].astype(np.float32).reshape(-1) for name in arr.dtype.names}
    else:
        height, width = arr.shape[0], arr.shape[1]
        bands = {"0": arr.astype(np.float32).reshape(-1)}
    return bands, width, height


def lng_lat_to_mercator_meters(lng: float, lat: float) -> tuple[float, float]:
    x = (lng * math.pi * EARTH_RADIUS_M) / 180
    clamped_lat = min(max(lat, -85.05112878), 85.05112878)
    y = EARTH_RADIUS_M * math.log(math.tan(math.pi / 4 + (clamped_lat * math.pi) / 360))
    return x, y


def mercator_meters_to_lng_lat(x: float, y: float) -> tuple[float, float]:
    lng = (x / (math.pi * EARTH_RADIUS_M)) * 180
    lat = ((2 * math.atan(math.exp(y / EARTH_RADIUS_M)) - math.pi / 2) * 180) / math.pi
    return lng, lat


# ── Segmentation rapide (seuillage + étiquetage + Voronoï, tout en C via scipy) ──


def watershed_segment(
    strength: np.ndarray,
    barrier: np.ndarray,
    width: int,
    height: int,
    seed_percentile: float = 0.2,
    min_seed_pixels: int = 6,
) -> np.ndarray:
    """
    strength : force de frontière par pixel (float), plus haut = plus proche d'une limite réelle.
    barrier  : True/1 = pixel non cultivable (route, eau, bâti...) : jamais inondé, agit comme séparateur.
    Retourne un tableau de labels (0 = non affecté/barrière, >=1 = identifiant de parcelle).

    Version rapide (C) : seuillage + étiquetage scipy + propagation Voronoï
    (plus proche germe). Remplace le tas binaire Python pur (~173 s/tuile 640²
    -> ~0.2 s) : les germes couvrent 20-30 % des pixels donc la frontière
    Voronoï est proche de la ligne de crête du gradient.
    """
    s = np.asarray(strength, dtype=np.float32).reshape(height, width)
    b = np.asarray(barrier, dtype=np.uint8).reshape(height, width)
    valid = s[b == 0]
    if valid.size == 0:
        return np.zeros((height * width,), dtype=np.int32)
    k = min(max(int(math.floor(valid.size * seed_percentile)), 0), valid.size - 1)
    threshold = float(np.partition(valid.ravel(), k)[k])

    seed2d = (s <= threshold) & (b == 0)
    if not bool(seed2d.any()):
        return np.zeros((height * width,), dtype=np.int32)
    try:
        from scipy.ndimage import distance_transform_edt as _edt
        from scipy.ndimage import label as _label
    except ImportError:
        _label = None  # type: ignore[assignment]
        _edt = None  # type: ignore[assignment]

    if _label is not None:
        struct = np.array([[0, 1, 0], [1, 1, 1], [0, 1, 0]], dtype=np.int32)
        comp, ncomp = _label(seed2d, structure=struct)
        if ncomp == 0:
            return np.zeros((height * width,), dtype=np.int32)
        counts = np.bincount(comp.ravel(), minlength=ncomp + 1)
        small = np.flatnonzero(counts < int(min_seed_pixels))
        if small.size:
            comp[np.isin(comp, small)] = 0
            # Recompacte pour des labels denses
            uniq = np.unique(comp)
            uniq = uniq[uniq > 0]
            if uniq.size == 0:
                return np.zeros((height * width,), dtype=np.int32)
            lut = np.zeros(int(comp.max()) + 1, dtype=np.int32)
            lut[uniq] = np.arange(1, uniq.size + 1)
            comp = lut[comp]
        if _edt is not None:
            # Voronoï : chaque pixel prend le label du germe le plus proche (C).
            _, idx = _edt(comp == 0, return_indices=True)
            out = comp[idx[0], idx[1]]
            out[b != 0] = 0
            return out.reshape(-1).astype(np.int32)
        # Sans EDT : retourne les germes seuls (mieux que rien).
        comp[b != 0] = 0
        return comp.reshape(-1).astype(np.int32)

    # Repli sans scipy (ne devrait plus arriver : scipy est une dépendance).
    labels = np.zeros((height, width), dtype=np.int32)
    labels[seed2d] = 1
    return labels.reshape(-1).astype(np.int32)


# ── Traçage de contours (Moore-neighbor tracing) + simplification Douglas-Peucker ──

Point = tuple[int, int]

_DIRECTIONS = (
    (1, 0),  # 0 E
    (1, 1),  # 1 SE
    (0, 1),  # 2 S
    (-1, 1),  # 3 SW
    (-1, 0),  # 4 W
    (-1, -1),  # 5 NW
    (0, -1),  # 6 N
    (1, -1),  # 7 NE
)


def _moore_trace(at, start_x: int, start_y: int, label: int, width: int, height: int) -> list[Point]:
    def is_foreground(x: int, y: int) -> bool:
        return at(x, y) == label

    boundary: list[Point] = [(start_x, start_y)]
    cx, cy = start_x, start_y
    # On "arrive" virtuellement depuis l'Ouest (le pixel de départ est le plus à
    # gauche de sa ligne, donc son voisin Ouest est nécessairement fond).
    backtrack_dir = 4
    max_steps = width * height * 8 + 8
    steps = 0

    while True:
        search_start = (backtrack_dir + 1) % 8
        found = False
        nx, ny, new_backtrack_dir = cx, cy, backtrack_dir
        for i in range(8):
            d = (search_start + i) % 8
            tx = cx + _DIRECTIONS[d][0]
            ty = cy + _DIRECTIONS[d][1]
            if is_foreground(tx, ty):
                nx, ny = tx, ty
                new_backtrack_dir = (d + 4) % 8
                found = True
                break
        if not found:
            break  # pixel isolé (pas de voisin foreground)

        cx, cy = nx, ny
        backtrack_dir = new_backtrack_dir
        steps += 1
        if cx == start_x and cy == start_y:
            break  # retour au point de départ : contour bouclé
        boundary.append((cx, cy))
        if steps >= max_steps:
            break

    return boundary


def trace_label_contours(labels: np.ndarray, width: int, height: int) -> dict[int, list[Point]]:
    """Trace le contour extérieur de chaque label présent dans le tableau, en
    coordonnées pixels (coins de pixels)."""
    contours: dict[int, list[Point]] = {}
    visited_start: set[int] = set()

    def at(x: int, y: int) -> int:
        if x < 0 or y < 0 or x >= width or y >= height:
            return 0
        return int(labels[y * width + x])

    for y in range(height):
        for x in range(width):
            label = at(x, y)
            if label <= 0:
                continue
            # Pixel de bord gauche d'une région : premier pixel de la ligne appartenant à ce
            # label, ou pixel dont le voisin de gauche a un label différent.
            if at(x - 1, y) == label:
                continue
            if label in visited_start:
                continue  # un seul contour externe tracé par label (composante principale)
            visited_start.add(label)

            contour = _moore_trace(at, x, y, label, width, height)
            if len(contour) >= 3:
                contours[label] = contour

    return contours


def _perpendicular_distance(p: tuple[float, float], a: tuple[float, float], b: tuple[float, float]) -> float:
    dx = b[0] - a[0]
    dy = b[1] - a[1]
    length_sq = dx * dx + dy * dy
    if length_sq == 0:
        return math.hypot(p[0] - a[0], p[1] - a[1])
    t = ((p[0] - a[0]) * dx + (p[1] - a[1]) * dy) / length_sq
    proj_x = a[0] + t * dx
    proj_y = a[1] + t * dy
    return math.hypot(p[0] - proj_x, p[1] - proj_y)


def _douglas_peucker(pts: list[tuple[float, float]], epsilon: float) -> list[tuple[float, float]]:
    if len(pts) <= 2:
        return pts
    max_dist = -1.0
    max_index = 0
    for i in range(1, len(pts) - 1):
        dist = _perpendicular_distance(pts[i], pts[0], pts[-1])
        if dist > max_dist:
            max_dist = dist
            max_index = i
    if max_dist > epsilon:
        left = _douglas_peucker(pts[: max_index + 1], epsilon)
        right = _douglas_peucker(pts[max_index:], epsilon)
        return left[:-1] + right
    return [pts[0], pts[-1]]


def simplify_polygon(points: list[tuple[float, float]], epsilon: float) -> list[tuple[float, float]]:
    if len(points) <= 2:
        return points
    return _douglas_peucker(points, epsilon)
